### ISSUE TITLE
`override` feature is unsafe or broken on many platforms and should be labeled as such

### ISSUE THREAD

**Issue by thomcc** [purpleprotocol/mimalloc_rust#41]
Mimalloc's MI_OVERRIDE feature is being used totally incorrectly for us. We always statically link, meaning it only works on a few unixes (never on windows), and that even then you must arrange for it to be first in the linkers path otherwise you'll end up 

I think it should be renamed to `unsafe_override` or `experimental_override` and disabled on platforms known to be incomplete (dragonflybsd) or buggy (macos), perhaps with an environment var you can use to force the issue.

**Comment by jq-rs** [comment]
New 1.7.0 mimalloc tries to fix DragonFly and MacOS part. Nevertheless, when I tried, it does not seem to work with rustc combined with libmimalloc-sys, at least. Any ideas what could be still missing?

**Comment by thomcc** [comment]
Yeah, I contributed some of the macOS fixes. The issue isn't mimalloc (necessarially), it's our use of it. The way to use malloc override on macOS is:

1. Build an override dylib with the override, zone, and interpose functionality enabled.
2. Launch your pogram with `DYLD_INSERT_LIBRARIES=path/to/libmimalloc.dylib` and `DYLD_FORCE_FLAT_NAMESPACE=1` variables set in the environment, e.g. `env DYLD_INSERT_LIBRARIES=libmimalloc.dylib DYLD_FORCE_FLAT_NAMESPACE=1 cargo build` or whatever.

`DYLD_FORCE_FLAT_NAMESPACE=1` changes how macOS's symbol resolution algorithm works. By default, macOS "fixes" symbol collisions by having them only apply to. E.g. if I link against two dylibs that both define a helper function with the same name, there's no collision on macOS (because the namespace is per-library) as there is on other OSes. However, when overriding malloc, we of course need it to be overridden in *all* locations. Overriding it just in one library be quite bad, as allocating memory in one place and freeing it in another is common. If you turn on `DYLD_FORCE_FLAT_NAMESPACE=1`, then there's only one namespace for symbols, and so the override applies globally.

Unfortunately, DYLD_FORCE_FLAT_NAMESPACE=1 tends to break things in strange ways. macOS system libraries don't do well with it on. This could be the issue you're seeing if you were doing the res to these steps properly. In general, there isn't a good way to override malloc globally and completely on mac for all programs.

So. You'll note that absolutely none of this is doable from the way we have the override feature setup, which it why I described it as being used totally incorrectly. The way we use it (to build a static library that is linked in) is broken on macOS, (and not just macOS either).

The only way to make this work without setting the environment is probably to detect those vars being missing in `main()` and if they are missing, set them and `execv` ourselves, which is highly highly dubious (actually, the DYLD_FORCE_FLAT_NAMESPACE one has a header in the Mach-O file format that turns it on automatically).

**Comment by thomcc** [comment]
For rustc on macOS my recommendation would be

1. Don't use any override functionality directly.
2. Statically link mimalloc into the program, and set a `#[global_allocator]` using it
3. Override C++'s `operator new`/`operator delete` from exactly one file. That can be done with https://github.com/microsoft/mimalloc/blob/master/include/mimalloc-new-delete.h in one of these: https://github.com/rust-lang/rust/tree/master/compiler/rustc_llvm/llvm-wrapper
4. Hope that LLVM doesn't assume `::operator new()` and `malloc` are the same allocator. A lot of programs do. They'd probably accept a patch for fixing any case where they do, though.

**Comment by jq-rs** [comment]
Thanks for the detailed response! Just wondering, jemalloc-sys in rustc claims that they can override malloc on MacOS. Of course, It may not really work? If it does, however, then maybe mimalloc can do it in a similar way. I'll take a look.

**Comment by thomcc** [comment]
It does not! https://github.com/gnzlbg/jemallocator/blob/master/jemalloc-sys/build.rs#L36-L45

**Comment by thomcc** [comment]
Note that `realpath` mentioned there the *least* of the issues, and truly just the tip of the iceberg when it comes to weird malloc override sadness on macOS.

**Comment by jq-rs** [comment]
Uhh, thanks. Rustc comments for config.toml are just wrong then.

**Comment by thomcc** [comment]
Huh, I had assumed they knew they weren't overriding on macOS. I'll bring this up in wg-compiler-performance.

**Comment by jq-rs** [comment]
And BTW, the reason I was looking into this is this test run: https://github.com/rust-lang/rust/pull/81782
Pretty good results, it looks like it would be easy to add a new mimalloc option to config.toml, as it seems MacOS can be ignored. The "override" is used here, seems to work fine for Linux, on proof-of-concept level anyway.

**Comment by thomcc** [comment]
> as it seems MacOS can be ignored

(*sad macOS user noises*)

But yes, switching rustc to use mimalloc is probably a good idea across the board.

**Comment by thomcc** [comment]
> The "override" is used here, seems to work fine for Linux, on proof-of-concept level anyway.

Also, note that this works iff `libmimalloc.a` is included first in the list of libraries given to the linker, anything before it will not necessarily use it. (It also is more likely to work if it's done in a single object file which we don't do right now but fairly easily could)

Since you're presumably adding this dep into the actual binary the way jemalloc does, that isn't a problem, but it's not exactly accurate that this means the `override` feature works without caveats.

**Comment by jq-rs** [comment]
Sounds to me that it would work on par with `jemalloc` at the moment. Maybe that is good enough, as it needs to be enabled from the config.toml explicitly. I can push the new option changes (as I already have them available) to the PR for reference and fix the confusing OSX part from the comments at the same time.

**Comment by thomcc** [comment]
> Sounds to me that it would work on par with jemalloc at the moment. Maybe that is good enough, as it needs to be enabled from the config.toml explicitly

Oh yes, it's fine for rustc, I just meant I still felt that it was worth renaming the feature to `unsafe_override` or something in this repository, so that any project that turns it on knows that some care is required.

**Comment by jq-rs** [comment]
@thomcc It seems possible to make override working for macOS with a reasonable effort, please see https://github.com/rust-lang/rust/issues/82423. Maybe this can be improved for `mimalloc` too (no sad macOS user after this :))

**Comment by thomcc** [comment]
In theory mimalloc supports it in specific cases, but not in a general/fully safe way, I'll look into what would be needed for rustc's case, but probably won't get to it until next week or so.

**Comment by jq-rs** [comment]
Thanks! I am not familiar with macOS internals, but as a macOS user can perhaps at least test drive.

**Comment by thomcc** [comment]
Do you have access to a pre-10.12 macOS machine?

**Comment by jq-rs** [comment]
I have the latest Big Sur only, so unfortunately no.

**Comment by thomcc** [comment]
That's fine, I'll see if I have a machine that's old enough (I think I have one in a box somewhere...). Sadly, a lot of these things changed then, and rustc supports back to 10.7

**Comment by madsmtm** [comment]
I believe this should be fixed by https://github.com/purpleprotocol/mimalloc_rust/pull/145 and https://github.com/purpleprotocol/mimalloc_rust/pull/146 - the first PR enables the dyld interposing support, the second ensures that the `static.o` file in `libmimalloc.a` is loaded by the linker.

Rust's version support has been bumped to macOS 10.12, so older versions not supporting dyld interposing shouldn't be a problem.

I have put up a repo in https://github.com/madsmtm/test-malloc-overriding for testing some of these things if you want to verify my analysis.
